src/Resolution.ts

```diff
@@ -453,6 +455,54 @@ export default class Resolution {
     return dnsUtils.toList(blockchainData);
   }
 
+  /**
+   * Retrieves the tokenURI from the registry smart contract.
+   * @returns the ERC721Metadata#tokenURI contract method result
+   * @param domain - domain name
+   */
+  async tokenURI(domain: string): Promise<string> {
+    const namehash = this.namehash(domain);
+    return this.getNamingMethodOrThrow(domain).getTokenUri(namehash);
+  }
+
+  /**
+   * Retrieves the data from the endpoint provided by tokenURI from the registry smart contract.
+   * @returns the JSON response of the token URI endpoint
+   * @param domain - domain name
+   */
+  async tokenURIMetadata(domain: string): Promise<TokenUriMetadata> {
+    const tokenUri = await this.tokenURI(domain);
+    return this.getMetadataFromTokenURI(tokenUri);
+  }
+
+  /**
+   * Retrieves the domain name from token metadata that is provided by tokenURI from the registry smart contract.
+   * The function will throw an error if the domain in the metadata does not match the hash (e.g. if the metadata is outdated).
+   * @returns the domain name retrieved from token metadata
+   * @param hash - domain hash
+   * @param service - nameservice which is used for lookup
+   */
+  async unhash(hash: string, service: NamingServiceName): Promise<string> {
+    const tokenUri = await this.serviceMap[service].getTokenUri(hash);
+    const metadata = await this.getMetadataFromTokenURI(tokenUri);
+    return metadata.name;
```

Comment:
As the service is centralized, you should validate the result locally (aka call `this.isValidHash`).

Reply:
Good catch, thanks! Forgot about the check.